openedx/core/djangoapps/user_api/preferences/tests/test_api.py

```diff
@@ -0,0 +1,324 @@
+
+    def test_get_user_preferences_errors(self):
+        """
+        Verifies that get_user_preferences returns appropriate errors.
+        """
+        with self.assertRaises(UserNotFound):
+            get_user_preferences(self.user, username="no_such_user")
+
+        with self.assertRaises(UserNotFound):
+            get_user_preferences(self.no_such_user)
+
+        with self.assertRaises(UserNotAuthorized):
+            get_user_preferences(self.different_user, username=self.user.username)
+
+    def test_set_user_preference(self):
+        """
+        Verifies the basic behavior of set_user_preference.
+        """
+        set_user_preference(self.user, self.test_preference_key, "new_value")
+        self.assertEqual(
+            get_user_preference(self.user, self.test_preference_key),
+            "new_value"
+        )
+
+    @patch('openedx.core.djangoapps.user_api.models.UserPreference.save')
+    def test_set_user_preference_errors(self, user_preference_save):
+        """
+        Verifies that set_user_preference returns appropriate errors.
+        """
+        with self.assertRaises(UserNotFound):
+            set_user_preference(self.user, self.test_preference_key, "new_value", username="no_such_user")
+
+        with self.assertRaises(UserNotFound):
+            set_user_preference(self.no_such_user, self.test_preference_key, "new_value")
+
+        with self.assertRaises(UserNotAuthorized):
+            set_user_preference(self.staff_user, self.test_preference_key, "new_value", username=self.user.username)
+
+        with self.assertRaises(UserNotAuthorized):
+            set_user_preference(self.different_user, self.test_preference_key, "new_value", username=self.user.username)
+
+        too_long_key = "x" * 256
+        with self.assertRaises(PreferenceValidationError):
+            set_user_preference(self.user, too_long_key, "new_value")
+
+        user_preference_save.side_effect = [Exception, None]
+        with self.assertRaises(PreferenceUpdateError):
+            set_user_preference(self.user, self.test_preference_key, "new_value")
+
+    def test_update_user_preferences(self):
+        """
+        Verifies the basic behavior of update_user_preferences.
+        """
+        expected_user_preferences = {
+            self.test_preference_key: "new_value",
+        }
+        update_user_preferences(self.user, expected_user_preferences), expected_user_preferences
+        self.assertEqual(
+            get_user_preference(self.user, self.test_preference_key),
+            "new_value"
+        )
+
+    @patch('openedx.core.djangoapps.user_api.models.UserPreference.delete')
+    @patch('openedx.core.djangoapps.user_api.models.UserPreference.save')
+    def test_update_user_preferences_errors(self, user_preference_save, user_preference_delete):
+        """
+        Verifies that set_user_preferences returns appropriate errors.
+        """
+        update_data = {
```

Comment:
Those cases are both tests in the views so I figured I wouldn't duplicate them all here. It is a tricky balancing act...